### ISSUE TITLE
Selective Editing blur radius different in preview

### ISSUE THREAD

**Issue by Lawrence37** [Beep6581/RawTherapee#7225]
**Short description**
When using Selective Editing > Blur/Grain & Denoise > Blur & Noise > Radius, the preview gives an inaccurate preview of the blur amount at zoom <100%.

**Steps to reproduce**
1. Enable Selective Editing, add a new spot, and add the Blur/Grain & Denoise tool.
2. Increase the blur radius.
3. Observe the blur amount at different zoom levels.

**Expected behavior**
The blur in the preview is close to the amount in the saved image.

Preview at 20% zoom:
![image](https://github.com/user-attachments/assets/b45ae428-a600-4ea6-992b-7c2c27d510f8)

Exported image, scaled down to 20%:
![image](https://github.com/user-attachments/assets/bfaeab8f-9d9c-4a7f-8e36-d831aea267c6)

Quick tests show that the blur amount is inversely proportional to the zoom, e.g. double the radius at 50%, triple the radius at 33%, quadruple the radius at 25%, etc.. It seems the radius is not scaled to compensate for the zoom factor.

**Comment by Desmis** [comment]
@Lawrence37 
This addition (blur) is now very old. I don't remember why I removed the Skip support. But for sure, the first version (well before 5.9, before LA appeared in dev) had "sk".

It's very easy to add, not even 2 minutes.

Jacques

**Comment by Lawrence37** [comment]
I discovered the bug is due to the formula being incorrect when using the Fourier transform convolution algorithm. This gives accurate blur at all preview sizes and fixes the difference in blur amount compared to the other algorithm.
```diff
diff --git a/rtengine/iplocallab.cc b/rtengine/iplocallab.cc
index 13f974d1a..f77839845 100644
--- a/rtengine/iplocallab.cc
+++ b/rtengine/iplocallab.cc
@@ -9523,8 +9523,8 @@ void ImProcFunctions::fftw_convol_blur(float * input, float * output, int bfw, i
 
     /*define the gaussian constants for the convolution kernel*/
     if (algo == 0) {
-        n_x = rtengine::RT_PI / (double) bfw; //ipol
-        n_y = rtengine::RT_PI / (double) bfh;
+        n_x = rtengine::RT_PI / (double) bfw / std::sqrt(2.); //ipol
+        n_y = rtengine::RT_PI / (double) bfh / std::sqrt(2.);
     } else if (algo == 1) {
         n_x = 1.f / bfw; //gauss
         n_y = 1.f / bfh;
@@ -9547,7 +9547,7 @@ void ImProcFunctions::fftw_convol_blur(float * input, float * output, int bfw, i
 
             for (int i = 0; i < bfw; i++)
                 if (algo == 0) {
-                    kern[ i + index] = exp((float)(-radius) * (n_x * i * i + n_y * j * j)); //calculate Gauss kernel Ipol formula
+                    kern[ i + index] = exp((float)(-radius * radius) * (n_x * i * i + n_y * j * j)); //calculate Gauss kernel Ipol formula
                 } else if (algo == 1) {
                     kern[ i + index] = radsig * exp((float)(-(n_x * i * i + n_y * j * j) / (2.f * radius * radius))); //calculate Gauss kernel  with Gauss formula
                 }
@@ -9585,7 +9585,7 @@ void ImProcFunctions::fftw_convol_blur(float * input, float * output, int bfw, i
                 int index = j * bfw;
 
                 for (int i = 0; i < bfw; i++) {
-                    out[i + index] *= exp((float)(-radius) * (n_x * i * i + n_y * j * j));    //apply Gauss kernel without FFT - some authors says radius*radius but differences with Gaussianblur
+                    out[i + index] *= exp((float)(-radius * radius) * (n_x * i * i + n_y * j * j));    //apply Gauss kernel without FFT - some authors says radius*radius but differences with Gaussianblur
                 }
             }
         } else if (algo == 1) {
```

The blur amount is drastically increased especially with high radii, so pp3 conversion and lowering the maximum radius would be nice. The new radius can be calculated from old pp3s using the formula `sqrt(2.8 * radius)`.

**Comment by Desmis** [comment]
@Lawrence37 
Thanks for this change. I just implemented it in the branch "fixbug-blur". I also changed the maximum limit from 10000 to 500 (since radius is now squared).
But this fft part is not a bug or an oversight, but a choice made 5 years ago. If you look at IPOL code and comments on FFT blur there is 2 ways : 
I join the texts between /* */ at the beginning of fftw_convol_blur
    
         Jacques Desmis june 2019 - inspired by Copyright 2013 IPOL Image Processing On Line http://www.ipol.im/
         when I read documentation on various FFT blur we found 2 possibilities
         0) kernel gauss is used with "normal" data
         1) kernel gauss is used with FFT
         fftkern allows to change 0) or 1) and test  It seems the good solution is with 0, but I keep the code in case of ??

         input real data to blur
         output real data blurred with radius
         bfw bfh width and high area
         radius = sigma for kernel
         n_x n_y relative width and high for kernel
         Gaussian blur is given by G(x,y) = (1/2*PI*sigma) * exp(-(x2 + y2) / 2* sigma2)
         its traduction in Fourier transform is G(x,y) =  exp((-sigma)*(PI * x2 + PI * y2)), for some authors it is not sigma but sigma^2..I have tried...huge differences with Gaussianblur
         after several test the only result that works very well is with fftkern = 0 and algo = 0, and as there is differences with Gaussianblur, I put an empirical correction in Ipretinex and Iplocalcontrast
         you can enabled or disabled this function with rtsettings.fftwsigma in options. By default empirical formula is disabled
         in fact no importance....if it is this function (for sigma) or another... we are not in research :)
    
       
**its traduction in Fourier transform is G(x,y) =  exp((-sigma)*(PI * x2 + PI * y2)), for some authors it is not sigma but sigma^2..I have tried...huge differences with Gaussianblur**


We had exchanges with Ingo and the Gaussianblur function (used in RT and which is an approximation) works badly above radius = 100, while the FFT function is (to the radius * radius, near) exact.

Nevertheless we can make the choice that you made. This is what I have just done without changing the pp3. We can do it later when we are in total agreement

I also changed the way the "Recovery based on luminance mask" function works. Now this function works with "Gaussian blur (noise and grain) , Median, and Guided Filter instead of only Guide Filter.

Jacques

**Comment by Lawrence37** [comment]
Where is the original code from IPOL? I looked in [A Survey of Gaussian Convolution Algorithms](https://www.ipol.im/pub/art/2013/87/article.pdf#subsection.3.2) and the [corresponding code](https://www.ipol.im/pub/art/2013/87/gaussian_20131215.tgz). Both of them use σ².
![image](https://github.com/user-attachments/assets/5ca8507b-65d2-4b28-9510-ce022c4477e0)

`gaussian_conv_dct.c`
```c
    temp = (sigma * M_PI) / width;
    c->dims.image.alpha_x = (num)(temp * temp / 2);
    temp = (sigma * M_PI) / height;
    c->dims.image.alpha_y = (num)(temp * temp / 2);
.
.
.
                    c.dest[x] *= ((num)exp(
                        -c.dims.image.alpha_x * x * x
                        -c.dims.image.alpha_y * y * y)) / denom;
```

**Comment by Lawrence37** [comment]
@Desmis I created a pull request (#7235) that also includes the pp3 value translations. There are several functions that use the FFTW blur, all of them in Selective Editing:
- Local contrast (unsharp mask method)
- Color & light mask
- Common color mask
- Color appearance mask
- Blur/Grain & denoise
- Dehaze and retinex

I believe the retinex values can't be changed without differences due to it having different radii used at different levels. Therefore, I kept the old behavior.
